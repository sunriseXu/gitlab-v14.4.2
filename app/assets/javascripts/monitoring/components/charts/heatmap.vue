<script>
import { GlHeatmap } from '@gitlab/ui/dist/charts';
import { formatDate, timezones, formats } from '../../format_date';
import { graphDataValidatorForValues } from '../../utils';

export default {
  components: {
    GlHeatmap,
  },
  props: {
    graphData: {
      type: Object,
      required: true,
      validator: graphDataValidatorForValues.bind(null, false),
    },
    timezone: {
      type: String,
      required: false,
      default: timezones.LOCAL,
    },
  },
  data() {
    return {
      width: 0,
    };
  },
  computed: {
    chartData() {
      return this.metrics.result.reduce(
        (acc, result, i) => [...acc, ...result.values.map((value, j) => [i, j, value[1]])],
        [],
      );
    },
    xAxisName() {
      return this.graphData.xLabel || '';
    },
    yAxisName() {
      return this.graphData.y_label || '';
    },
    xAxisLabels() {
      return this.metrics.result.map((res) => Object.values(res.metric)[0]);
    },
    yAxisLabels() {
      return this.result.values.map((val) => {
        const [yLabel] = val;

        return formatDate(new Date(yLabel), {
          format: formats.shortTime,
          timezone: this.timezone,
        });
      });
    },
    result() {
      return this.metrics.result[0];
    },
    metrics() {
      return this.graphData.metrics[0];
    },
  },
};
</script>
<template>
  <gl-heatmap
    ref="heatmapChart"
    v-bind="$attrs"
    :responsive="true"
    :data-series="chartData"
    :x-axis-name="xAxisName"
    :y-axis-name="yAxisName"
    :x-axis-labels="xAxisLabels"
    :y-axis-labels="yAxisLabels"
    :width="width"
  />
</template>
