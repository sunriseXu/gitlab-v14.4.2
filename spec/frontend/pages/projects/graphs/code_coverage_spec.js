import { GlAlert, GlDropdown, GlDropdownItem } from '@gitlab/ui';
import { GlAreaChart } from '@gitlab/ui/dist/charts';
import { shallowMount } from '@vue/test-utils';
import MockAdapter from 'axios-mock-adapter';

import { nextTick } from 'vue';
import waitForPromises from 'helpers/wait_for_promises';
import axios from '~/lib/utils/axios_utils';
import httpStatusCodes from '~/lib/utils/http_status';
import CodeCoverage from '~/pages/projects/graphs/components/code_coverage.vue';
import { codeCoverageMockData, sortedDataByDates } from './mock_data';

describe('Code Coverage', () => {
  let wrapper;
  let mockAxios;

  const graphEndpoint = '/graph';
  const graphStartDate = '13 February';
  const graphEndDate = '12 May';
  const graphRef = 'master';
  const graphCsvPath = 'url/';

  const findAlert = () => wrapper.findComponent(GlAlert);
  const findAreaChart = () => wrapper.findComponent(GlAreaChart);
  const findAllDropdownItems = () => wrapper.findAllComponents(GlDropdownItem);
  const findFirstDropdownItem = () => findAllDropdownItems().at(0);
  const findSecondDropdownItem = () => findAllDropdownItems().at(1);
  const findDownloadButton = () => wrapper.find('[data-testid="download-button"]');

  const createComponent = () => {
    wrapper = shallowMount(CodeCoverage, {
      propsData: {
        graphEndpoint,
        graphStartDate,
        graphEndDate,
        graphRef,
        graphCsvPath,
      },
    });
  };

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  describe('when fetching data is successful', () => {
    beforeEach(() => {
      mockAxios = new MockAdapter(axios);
      mockAxios.onGet().replyOnce(httpStatusCodes.OK, codeCoverageMockData);

      createComponent();

      return waitForPromises();
    });

    afterEach(() => {
      mockAxios.restore();
    });

    it('renders the area chart', () => {
      expect(findAreaChart().exists()).toBe(true);
    });

    it('sorts the dates in ascending order', () => {
      expect(wrapper.vm.sortedData).toEqual(sortedDataByDates);
    });

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('shows no error messages', () => {
      expect(findAlert().exists()).toBe(false);
    });

    it('does not render download button', () => {
      expect(findDownloadButton().exists()).toBe(true);
    });
  });

  describe('when fetching data fails', () => {
    beforeEach(() => {
      mockAxios = new MockAdapter(axios);
      mockAxios.onGet().replyOnce(httpStatusCodes.BAD_REQUEST);

      createComponent();

      return waitForPromises();
    });

    afterEach(() => {
      mockAxios.restore();
    });

    it('renders an error message', () => {
      expect(findAlert().exists()).toBe(true);
      expect(findAlert().attributes().variant).toBe('danger');
    });

    it('still renders an empty graph', () => {
      expect(findAreaChart().exists()).toBe(true);
    });
  });

  describe('when fetching data succeed but returns an empty state', () => {
    beforeEach(() => {
      mockAxios = new MockAdapter(axios);
      mockAxios.onGet().replyOnce(httpStatusCodes.OK, []);

      createComponent();

      return waitForPromises();
    });

    afterEach(() => {
      mockAxios.restore();
    });

    it('renders an information message', () => {
      expect(findAlert().exists()).toBe(true);
      expect(findAlert().attributes().variant).toBe('info');
    });

    it('still renders an empty graph', () => {
      expect(findAreaChart().exists()).toBe(true);
    });

    it('does not render download button', () => {
      expect(findDownloadButton().exists()).toBe(false);
    });
  });

  describe('dropdown options', () => {
    beforeEach(() => {
      mockAxios = new MockAdapter(axios);
      mockAxios.onGet().replyOnce(httpStatusCodes.OK, codeCoverageMockData);

      createComponent();

      return waitForPromises();
    });

    it('renders the dropdown with all custom names as options', () => {
      expect(wrapper.findComponent(GlDropdown).exists()).toBeDefined();
      expect(findAllDropdownItems()).toHaveLength(codeCoverageMockData.length);
      expect(findFirstDropdownItem().text()).toBe(codeCoverageMockData[0].group_name);
    });
  });

  describe('interactions', () => {
    beforeEach(() => {
      mockAxios = new MockAdapter(axios);
      mockAxios.onGet().replyOnce(httpStatusCodes.OK, codeCoverageMockData);

      createComponent();

      return waitForPromises();
    });

    it('updates the selected dropdown option with an icon', async () => {
      findSecondDropdownItem().vm.$emit('click');

      await nextTick();

      expect(findFirstDropdownItem().attributes('ischecked')).toBe(undefined);
      expect(findSecondDropdownItem().attributes('ischecked')).toBe('true');
    });

    it('updates the graph data when selecting a different option in dropdown', async () => {
      const originalSelectedData = wrapper.vm.selectedDailyCoverage;
      const expectedData = codeCoverageMockData[1];

      findSecondDropdownItem().vm.$emit('click');

      await nextTick();

      expect(wrapper.vm.selectedDailyCoverage).not.toBe(originalSelectedData);
      expect(wrapper.vm.selectedDailyCoverage).toBe(expectedData);
    });
  });
});
