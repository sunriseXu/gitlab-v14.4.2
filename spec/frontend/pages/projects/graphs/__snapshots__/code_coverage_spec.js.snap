// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code Coverage when fetching data is successful matches the snapshot 1`] = `
<div>
  <div
    class="gl-display-flex gl-justify-content-space-between gl-align-items-center gl-border-t gl-pt-4 gl-mb-3"
  >
    <h4
      class="gl-m-0"
      sub-header=""
    >
      <gl-sprintf-stub
        message="Code coverage statistics for %{ref} %{start_date} - %{end_date}"
      />
    </h4>
     
    <gl-button-stub
      buttontextclasses=""
      category="primary"
      data-testid="download-button"
      href="url/"
      icon=""
      size="small"
      variant="default"
    >
      
      Download raw data (.csv)
    
    </gl-button-stub>
  </div>
   
  <div
    class="gl-mt-3 gl-mb-3"
  >
    <!---->
     
    <!---->
     
    <gl-dropdown-stub
      category="primary"
      clearalltext="Clear all"
      clearalltextclass="gl-px-5"
      headertext=""
      hideheaderborder="true"
      highlighteditemstitle="Selected"
      highlighteditemstitleclass="gl-px-5"
      size="medium"
      text="rspec"
      variant="default"
    >
      <gl-dropdown-item-stub
        avatarurl=""
        iconcolor=""
        iconname=""
        iconrightarialabel=""
        iconrightname=""
        ischecked="true"
        ischeckitem="true"
        secondarytext=""
        value="rspec"
      >
        
        rspec
      
      </gl-dropdown-item-stub>
      <gl-dropdown-item-stub
        avatarurl=""
        iconcolor=""
        iconname=""
        iconrightarialabel=""
        iconrightname=""
        ischeckitem="true"
        secondarytext=""
        value="cypress"
      >
        
        cypress
      
      </gl-dropdown-item-stub>
      <gl-dropdown-item-stub
        avatarurl=""
        iconcolor=""
        iconname=""
        iconrightarialabel=""
        iconrightname=""
        ischeckitem="true"
        secondarytext=""
        value="karma"
      >
        
        karma
      
      </gl-dropdown-item-stub>
    </gl-dropdown-stub>
  </div>
   
  <gl-area-chart-stub
    annotations=""
    data="[object Object]"
    formattooltiptext="[Function]"
    height="200"
    includelegendavgmax="true"
    legendaveragetext="Avg"
    legendcurrenttext="Current"
    legendlayout="inline"
    legendmaxtext="Max"
    legendmintext="Min"
    option="[object Object]"
    responsive=""
    thresholds=""
  />
</div>
`;
